Allow peers to resync up to consensus

jkitman opened the pull request:
Allows peers to resync up to consensus
   - Upon start-up (possibly from crash) we send `RejoinRequest` to all peers
   - Peers send the HBBFT `next_epoch` back to us and last signed epoch history
   - We validate the epoch history sigs and that epoch history is within `f + 3` of the `next_epoch`
   - Take `max(next_epoch)` from `2f + 1` peers an run HBBFT to produce an outcome
   - Download all the epochs we've missed and process them as we can validate them

elsirion commented on minimint/src/lib.rs:
```diff
@@ -145,6 +173,89 @@ impl MinimintServer {
+
+        // Verify signature if we downloaded any epochs
+        if let Some(mut prev_epoch) = prev_epoch {
+            assert!(end_epoch
```
The signature being wrong should probably be an error, we might handle it by querying a previous peer.

elsirion commented on minimint/src/lib.rs:
```diff
@@ -145,6 +173,89 @@ impl MinimintServer {
+        // Verify signature if we downloaded any epochs
+        if let Some(mut prev_epoch) = prev_epoch {
+            assert!(end_epoch
+                .add_sig_to_prev(&self.cfg.epoch_pk_set, &mut prev_epoch)
```
We should have
1. more documentation why this is secure
2. maybe do some double-checking (check both signatures and hashes?)

to make it easier to understand why this is secure. I figured it out after a while, but the commitment structure and where checks happen isn't obvious at all.

elsirion commented on minimint/src/lib.rs:
```diff
@@ -145,6 +173,89 @@ impl MinimintServer {
+                "Rejoining consensus: processing epoch {}",
+                epoch.outcome.epoch
+            );
+            let outcome = ConsensusOutcomeConversion::from(epoch.outcome).0;
+            self.consensus.process_consensus_outcome(outcome).await;
```
Do we need a struct for the conversion here? Wouldn't a direct impl of `From` work? Or just a plain old function?

jkitman replied:
Once we can get remove `ConsensusOutcomes` dependency on hbbft then we can eliminate this.

elsirion commented on minimint/src/lib.rs:
```diff
@@ -209,24 +320,43 @@ impl MinimintServer {
         }
     }
 
+    fn start_next_epoch(&self, msg: &PeerMessage) -> bool {
```
We might want to run a few epochs here, e.g. at least 1 every 10s for 100s (more if there are actual items) to allow the peer to join with some safety margin (they might want to join `next_epoch+5`).

elsirion commented on minimint/src/db.rs:
```diff
@@ -67,7 +67,7 @@ impl DatabaseKeyPrefixConst for DropPeerKeyPrefix {
 }
 
-#[derive(Debug, Encodable, Decodable)]
+#[derive(Debug, Clone, Encodable, Decodable)]
```
This could also be `Copy`

elsirion commented on minimint/Cargo.toml:
```diff
@@ -33,6 +33,7 @@ hbbft = { git = "https://github.com/fedimint/hbbft" }
 hex = "0.4.2"
 itertools = "0.10.0"
 jsonrpsee = { version = "0.14.0", features = ["ws-server"] }
+mint-client = { path = "../client/client-lib" }
```
Ideally we could move the `FederationApi` trait somewhere else … (e.g. core)

jkitman replied:
Will tackle separately.

elsirion reviewed (changes requested):
Great progress! I left a lot of nits and a few comments that might be security-relevant.

jkitman commented on minimint/src/consensus/mod.rs:
```diff
@@ -321,56 +321,34 @@ where
         &self,
         outcome: ConsensusOutcome,
-        signatures: Vec<(PeerId, EpochSignatureShare)>,
+        _signatures: Vec<(PeerId, EpochSignatureShare)>,
```
Yep a refactor made that unnecessary.

jkitman commented on minimint/src/consensus/mod.rs:
```diff
@@ -321,56 +321,34 @@ where
-                transaction.append_insert(prev_epoch_key, prev_epoch);
-            } else {
-                warn!("Unable to sign epoch {}", prev_epoch.outcome.epoch);
+                Err(_) => warn!("Unable to sign epoch {}", prev_epoch.outcome.epoch),
```
Good catch, I missed that when I refactored!

jkitman commented on minimint/src/lib.rs:
```diff
@@ -209,24 +320,43 @@ impl MinimintServer {
-            .handle_message(&peer, peer_msg)
-            .expect("HBBFT handle message failed");
+        match msg {
+            (_, EpochMessage::RejoinEpoch(_)) => vec![],
```
It's possible that once we receive enough messages to know the `next_epoch` we are still receiving and ignore the remaining messages.

jkitman commented on minimint/src/lib.rs:
```diff
@@ -145,6 +173,89 @@ impl MinimintServer {
+                (peer, EpochMessage::RejoinEpoch(epoch)) => {
+                    next_epochs.insert(peer, epoch);
+                }
+                msg if next_epochs.contains_key(&msg.0) => msg_buffer.push(msg),
```
Peers could have old messages in the send buffer.

But now that I think about it, it might not be safe to drop them in case the new epoch starts before a peer responds to the `RejoinRequest`

jkitman commented on minimint/src/lib.rs:
```diff
@@ -209,24 +320,43 @@ impl MinimintServer {
-            .expect("HBBFT handle message failed");
+        match msg {
+            (_, EpochMessage::RejoinEpoch(_)) => vec![],
+            (peer, EpochMessage::RejoinRequest) => {
+                let target = Target::Nodes(BTreeSet::from([peer]));
+                let next_epoch = EpochMessage::RejoinEpoch(self.hbbft.next_epoch());
+                self.connections.send(target, next_epoch).await;
+                vec![]
+            }
```
My plan is to handle the send buffers in a different PR.  For now rejoining will work but the send buffers will eventually block consensus.

jkitman commented on minimint/src/lib.rs:
```diff
@@ -145,6 +173,89 @@ impl MinimintServer {
+            )
+            .await;
+
+        // Advance consensus to the max(next_epoch) from max_evil + 1 peers
```
I was hoping to avoid something like a safety margin since it's kind of hacky.

I think what I'll do instead is something I mentioned earlier - submit the rejoin request as a consensus item.  Then all the honest peers can flush their send buffers and begin sending hbbft messages at the next epoch.

jkitman commented on minimint/src/lib.rs:
```diff
@@ -145,6 +173,89 @@ impl MinimintServer {
+        // Run consensus until arriving at the next_epoch consensus outcome
+        let mut outcomes: Vec<ConsensusOutcome> = vec![];
+        let mut next_outcome = None;
+
+        for msg in msg_buffer {
+            outcomes.append(&mut self.handle_message(msg).await);
+        }
+
+        while next_outcome.is_none() {
+            let msg = self.connections.receive().await;
+            outcomes.append(&mut self.handle_message(msg).await);
+            next_outcome = outcomes.iter().find(|out| out.epoch == next_epoch_num);
+        }
+        tracing::info!("Rejoining consensus: created outcome");
```
We could, although we wouldn't know when to stop the download.

jkitman commented on minimint/src/lib.rs:
```diff
@@ -145,6 +173,89 @@ impl MinimintServer {
+            let epoch = self.api.fetch_epoch_history(i).await.unwrap();
+            assert!(epoch.verify_hash(&prev_epoch).is_ok());
+            prev_epoch = Some(epoch.clone());
+            epochs.push(epoch);
```
We could process epochs as soon as we get a valid sig in that case.

jkitman commented on minimint/src/lib.rs:
```diff
@@ -145,6 +173,89 @@ impl MinimintServer {
+        let mut next_epochs = BTreeMap::<PeerId, u64>::new();
+
+        tracing::info!("Rejoining consensus");
+        while next_epochs.len() < self.cfg.max_faulty() + 1 {
```
The median might not be safe, for instance if `p` peers are on `epoch <= n` with `(3f + 1) / 2 < p < 2f + 1` and the remaining peers are on `epoch > n + buffer`.

For instance with `7` peers we could have `4` peers on epoch `1` and `3` peers on epoch `6` with a buffer of `5`.  The rejoining peer would never receive enough messages to determine consensus.

jkitman replied:
An alternative algorithm that should work is we require `2f + 1` peers to respond with the next epoch and include the last signed consensus history so we can validate the signatures (confirming we can download all prior history).  We then take `max(valid responses)`.

The other nice benefit to this solution is it doesn't require a buffer and running empty epochs which feels a bit hacky and error-prone.

elsirion replied:
That sounds like a great idea now that we have signed epochs! :smiley:

jkitman commented:
@elsirion @justinmoon This PR is ready for another review, I've incorporated most of your feedback and added a bash test to confirm it actually works on a real federation.

There are still several tasks left before peer reconnection is robust which I've documented here: https://github.com/fedimint/fedimint/issues/320 but I wanted to get this PR in first since it's already large and can stand on its own I think.

elsirion reviewed (commented):
I think the logic is sound, which is no small feat! This is one of the most complicated parts. I left some nits that are mostly about structure/clarity so the code is easier to comprehend.

jkitman commented on minimint/src/lib.rs:
```diff
@@ -145,6 +173,89 @@ impl MinimintServer {
+            };
+        }
+        let next_epoch_num = next_epochs.into_values().max().unwrap();
+        self.hbbft.skip_to_epoch(next_epoch_num);
```
So the cool thing about using the `2f+1` max of `hbbft.next_epoch` is that we don't need to worry about latency since the `Rejoin` response occurs within the hbbft messaging thread, assuring no hbbft messages have been sent yet from that epoch.

jkitman commented on minimint/src/lib.rs:
```diff
@@ -145,6 +173,89 @@ impl MinimintServer {
+
+        for i in download_epoch_num..next_epoch_num {
+            let epoch = self.api.fetch_epoch_history(i).await.unwrap();
+            assert!(epoch.verify_hash(&prev_epoch).is_ok());
```
Will need to handle in a separate PR once we have the API query strategies.

jkitman commented on scripts/reconnect-test.sh:
```diff
@@ -0,0 +1,29 @@
+await_block_sync
+
+./scripts/start-fed.sh
+sleep 5
```
The problem with that is again due to the lack of an API strategy, any polling will wait until just 3 peers respond, but we'd want to wait until 4.  So this is kind of hacky until we can configure that.  I can add a `FIXME` note.

jkitman commented on scripts/final-checks.sh:
```diff
@@ -6,10 +6,11 @@
 
 export FM_TEST_DISABLE_MOCKS=0
-cargo test
+cargo +nightly test --release
```
When I don't use `+nightly` I get a compile failure due to unstable `rocksdb_impl`.

dpc replied:
This was the comment I missed and couldn't find. Will check.

dpc replied:
@jkitman It works here. Please ping me on discord to debug it. Possibly it's `cargo-clippy` in your `$HOME/.cargo/bin`?

dpc replied:
In the meantime: https://github.com/fedimint/fedimint/pull/490

jkitman commented on fedimint/src/net/peers.rs:
```diff
@@ -421,10 +424,10 @@ where
             .last_received
             .map(|last_id| last_id.increment())
-            .unwrap_or(MessageId(1));
+            .unwrap_or(msg.id);
```
If the peer restarts then they forget what message they are on, but its peers do not.  Perhaps once the buffer drop is implemented then we can change this back.

jkitman commented on fedimint-core/src/epoch.rs:
```diff
@@ -71,12 +71,42 @@ impl EpochHistory {
-        pks: &PublicKey,
-        prev_epoch: Option<&EpochHistory>,
+        pks: &PublicKeySet,
+        prev_epoch: &mut EpochHistory,
+        contributing_peers: &mut HashSet<PeerId>,
```
Yeah I'm not a fan of this pattern either.  The thing about `contributing_peers` is I need to return in it the error case as well as the success case which seems odd, but I can make that change.

jkitman commented on fedimint/src/lib.rs:
```diff
@@ -145,6 +177,102 @@ impl FedimintServer {
+        }
+        tracing::info!("Rejoining consensus: created outcome");
+        // FIXME: Should handle failing and querying other peers
+        self.download_history(outcomes[0].clone()).await.unwrap();
```
Actually it's possible that we create epochs from the near-future.  This was another possible solution to the resync I tried to explain on our call, but I prefer the custom messages solution in this PR because it'll be more useful when I add the handling of a consensus restart scenario.

This is actually an assumption that occurs in our normal hbbft handling code (that epochs are received in-order) and should be tracked as a separate issue (I'll create one for that now).

jkitman commented on fedimint/src/lib.rs:
```diff
@@ -209,24 +337,59 @@ impl FedimintServer {
+    fn start_next_epoch(&self, msg: &PeerMessage) -> bool {
+        match msg {
+            (_, EpochMessage::Continue(peer_msg)) => self.hbbft.next_epoch() == peer_msg.epoch(),
+            (_, EpochMessage::RejoinRequest) => true,
```
So the 3 epochs later is only setting an upper bound on what the epoch will be, but in most scenarios the fed is probably idle awaiting new transactions in which case it only will take 1 epoch for the rejoining peer to catch up (this is demonstrated by the tests).
